File: FSharp/Algorithms/Sauvola.fs
namespace Functional

open System

module Sauvola =
    let CalculateIntegralImage(pixels: byte[], width: int, height: int, transform: bool) =
        let integralImage = Array2D.zeroCreate<int> height width

        for y = 0 to height - 1 do
            for x = 0 to width - 1 do
                let mutable pixelValue = int pixels.[y * width + x]
                if transform then pixelValue <- pixelValue * pixelValue
                integralImage.[y, x] <- pixelValue
                if x > 0 then integralImage.[y, x] <- integralImage.[y, x] + integralImage.[y, x - 1]
                if y > 0 then integralImage.[y, x] <- integralImage.[y, x] + integralImage.[y - 1, x]
                if x > 0 && y > 0 then integralImage.[y, x] <- integralImage.[y, x] - integralImage.[y - 1, x - 1]

        integralImage

    let MeanStdIntegralImage(pixels: byte[], width: int, height: int, windowSize: int) =
        let integralImage = CalculateIntegralImage(pixels, width, height, false)
        let squaredIntegralImage = CalculateIntegralImage(pixels, width, height, true)

        let mean = Array.zeroCreate<float>(width * height)
        let std = Array.zeroCreate<float>(width * height)
        let half = windowSize / 2
        let windowArea = windowSize * windowSize

        for y = 0 to height - 1 do
            for x = 0 to width - 1 do
                let x1 = max (x - half) 0
                let y1 = max (y - half) 0
                let x2 = min (x + half) (width - 1)
                let y2 = min (y + half) (height - 1)

                let sum = integralImage.[y2, x2] - integralImage.[y1, x2] - integralImage.[y2, x1] + integralImage.[y1, x1]
                let sumSquares = squaredIntegralImage.[y2, x2] - squaredIntegralImage.[y1, x2] - squaredIntegralImage.[y2, x1] + squaredIntegralImage.[y1, x1]

                let meanValue = float sum / float windowArea
                let variance = (float sumSquares / float windowArea) - (meanValue * meanValue)
                let stdValue = Math.Sqrt variance

                let i = y * width + x
                mean.[i] <- meanValue
                std.[i] <- stdValue

        (mean, std)

    let Binarize(pixels: byte[], width: int, height: int) =
        let windowSize = int (float width * 0.02)
        let k = 0.2
        let r = 127.5

        let (mean, std) = MeanStdIntegralImage(pixels, width, height, windowSize)

        for y = 0 to height - 1 do
            for x = 0 to width - 1 do
                let i = y * width + x
                let threshold = mean.[i] * (1.0 + k * (std.[i] / r - 1.0))
                pixels.[i] <- if pixels.[i] > byte threshold then 255uy else 0uy

        pixels

File: CSharp/Algorithms/Declarative/Sauvola.cs
namespace CSharp.Algorithms.Declarative;

public class Sauvola : IBinarizationAlgorithm
{
    public string Name => "Sauvola";
    public string Paradigm => "Declarative";

    public void Binarize(byte[] pixels, int width, int height)
    {
        int windowSize = (int)(width * 0.02f);
        const float k = 0.2f;
        const float r = 127.5f;

        var (mean, std) = MeanStdIntegralImage(pixels, width, height, windowSize);

        var transformedPixels = pixels.Select((pixel, index) =>
                pixel > mean[index] * (1 + k * (std[index] / r - 1))
                    ? byte.MaxValue
                    : byte.MinValue)
            .ToArray();

        Array.Copy(transformedPixels, pixels, pixels.Length);
    }

    private static (float[], float[]) MeanStdIntegralImage(byte[] pixels, int width, int height, int windowSize)
    {
        int[,] integralImage = CalculateIntegralImage(pixels, width, height);
        int[,] squaredIntegralImage = CalculateIntegralImage(pixels, width, height, pixel => pixel * pixel);

        var mean = new float[width * height];
        var std = new float[width * height];
        var half = windowSize / 2;
        var windowArea = windowSize * windowSize;

        for (var y = 0; y < height; y++)
        {
            for (var x = 0; x < width; x++)
            {
                int x1 = Math.Max(x - half, 0);
                int y1 = Math.Max(y - half, 0);
                int x2 = Math.Min(x + half, width - 1);
                int y2 = Math.Min(y + half, height - 1);

                int sum = integralImage[y2, x2] - integralImage[y1, x2] - integralImage[y2, x1] + integralImage[y1, x1];
                int sumSquares = squaredIntegralImage[y2, x2] - squaredIntegralImage[y1, x2] - squaredIntegralImage[y2, x1] + squaredIntegralImage[y1, x1];

                float meanValue = (float)sum / windowArea;
                float variance = (sumSquares / windowArea) - (meanValue * meanValue);
                float stdValue = (float)Math.Sqrt(variance);

                var i = y * width + x;
                mean[i] = meanValue;
                std[i] = stdValue;
            }
        }

        return (mean, std);
    }

    private static int[,] CalculateIntegralImage(byte[] pixels, int width, int height, Func<byte, int> transform = null)
    {
        var integralImage = new int[height, width];

        for (int y = 0; y < height; y++)
        {
            for (int x = 0; x < width; x++)
            {
                int pixelValue = (transform != null) ? transform(pixels[y * width + x]) : pixels[y * width + x];
                integralImage[y, x] = pixelValue;
                if (x > 0) integralImage[y, x] += integralImage[y, x - 1];
                if (y > 0) integralImage[y, x] += integralImage[y - 1, x];
                if (x > 0 && y > 0) integralImage[y, x] -= integralImage[y - 1, x - 1];
            }
        }

        return integralImage;
    }
}